### ISSUE TITLE
mpz_get_si() behaviour

### ISSUE THREAD

**Issue by mkskeller** [wbhart/mpir#173]
mpz_get_si() doesn't preserve the sign any more in some cases.

```
  mpz_class a = 1;
  a <<= 63;
  cout << a << " " << a.get_si() << endl;
```

Above code outputs 
`9223372036854775808 -9223372036854775808`
but in version 2.5 it used to output
`9223372036854775808 0`

**Comment by wbhart** [comment]
Is the behaviour defined for mpz_get_si if the result doesn't fit into a
signed word?

**Comment by mkskeller** [comment]
The documentation says: "If op fits into a mpir_si return the value of 
op. Otherwise return the least significant part of op, with the same 
sign as op."

Am 17.09.2016 um 10:16 schrieb wbhart:

**Comment by wbhart** [comment]
Thanks for reporting this. It is clearly a bug.

Are you encountering this on Windows or Linux? And just to confirm: you
found this with version 2.7.2 of MPIR?

Bill.

**Comment by mkskeller** [comment]
On Linux, both with 2.7.2 and the master branch from Github. The 
attached patch should fix it, for positive and negative numbers.

Marcel

**Comment by mkskeller** [comment]
From 43ead0838f69c49b22f8f72ae3e54507d679a256 Mon Sep 17 00:00:00 2001
From: Marcel Keller <m.keller@bristol.ac.uk>
Date: Mon, 19 Sep 2016 11:43:54 +0100
Subject: [PATCH] Preserve sign when converting.

---
 mpz/get_si.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mpz/get_si.c b/mpz/get_si.c
index ae7b298..65662f3 100644
--- a/mpz/get_si.c
+++ b/mpz/get_si.c
@@ -36,10 +36,10 @@ mpz_get_si (mpz_srcptr z)
 #endif
 
   if (size > 0)
-    return (mpir_si) zl & GMP_UI_MAX;
+    return (mpir_si) zl & GMP_SI_MAX;
   else if (size < 0)
     /* This expression is necessary to properly handle 0x80000000 */
-    return ~(((mpir_si) zl - 1L) & GMP_UI_MAX);
+    return ~(((mpir_si) zl - 1L) & GMP_SI_MAX);
   else
     return 0;
 }
-- 
2.9.3

**Comment by mkskeller** [comment]
Patch lost, made a pull request instead.

**Comment by tholberton** [comment]
@wbhart looks like this issue was fixed in 622672e.

**Comment by wbhart** [comment]
Indeed.
